source/panning/VirtualCinematographer.cpp

```diff
@@ -29,6 +29,80 @@ using namespace std;
 
 int skipLecturePosition = 2; //Only process every second point of lecturers position (Helps filter noisy movement/jitter out)
 
+
+void DefaultVirtualCinematographerOutput::outputHeader(ofstream & stream, const PersistentData & persistentData) {
+    stream << "# track4k " << persistentData.inputFile << " " << persistentData.processedFrames
+             << " frames (frame top-left-x top-left-y) output frame size "
+             << persistentData.panOutputVideoSize.width << " " << persistentData.panOutputVideoSize.height << endl;
+}
+
+void DefaultVirtualCinematographerOutput::outputFrames(ofstream & stream, const PersistentData & persistentData, const std::vector<Rect> & cropRectangles, long int y) {
+     int last_x = -1;
+     int i;
+
+    // Write out the pan x position and the fixed y position
+    for (i = 0; i < persistentData.processedFrames - 1; i++) {
+        if (cropRectangles[i].x != last_x) {
+            stream << i << " " << cropRectangles[i].x << " " << y << endl;
+            last_x = cropRectangles[i].x;
+        }
+    }
+
+    // Always write out the last frame
+    stream << i << " " << cropRectangles[i].x << " " << y << endl;
+}
+
+void JsonVirtualCinematographerOutput::outputHeader(ofstream & stream, const PersistentData & persistentData) {
+    stream << "{" << endl;
+    stream << "\t" << "\"inputFile\":\"" << persistentData.inputFile <<
+        "\",\n\t\"originalWidth\":" << persistentData.videoDimension.width <<
+        ", \"originalHeight\":" << persistentData.videoDimension.height <<
+        ",\n\t\"processedFrames\": " << persistentData.processedFrames <<
+        ",\n\t\"width\": " << persistentData.panOutputVideoSize.width << ", \"height\": " << persistentData.panOutputVideoSize.height << "," << std::endl;
+}
+
+void JsonVirtualCinematographerOutput::outputFrames(ofstream & stream, const PersistentData & persistentData, const std::vector<Rect> & cropRectangles, long int y) {
+    int last_x = -1;
+    int i;
+    double fps = persistentData.fps;
+    int lastSecond = -1;
+
+    auto printFrame = [&](int frame, int cropX, long int cropY,bool printComma) {
+        int second = static_cast<int>(static_cast<double>(frame) / fps);
```

Comment:
This time value in seconds will not always be correct when the input video has variable frame rate (for example from cameras that produce an encoded RTSP H264 stream, like Axis 1428). The fps value is also an approximation from OpenCV (duration / frames).

So it would be preferable here to include the frame count as well (frame: XXX), and for the timestamp, include millisecond precision (time: XXX.YYY).

Regardless of whether paella will use the frame count or not, it's better to be accurate (for other potential applications) and so that the json is a superset of the txt output and not losing information.

Reply:
Also depending on the container, OpenCV may not even return an accurate frame count. It seems to actual have the correct duration and return a calculated number of frames, but the actual number of frames processed can differ so the fps may just be wrong.

This was the motivation for splitting track4k into track4k and cropvid, as OpenCV doesn't provide accurate per-frame timing information, but we can get that with ffmpeg. So the only accurate item in the OpenCV output is the frame number.

To get accurate timestamps in the json, we'd have to re-process it with the source video using ffmpeg libraries or similar to get the per-frame PTS (similar to what cropvid does, just updating a timestamp rather than creating an output video).